src/ann_solo/utils.py

```diff
@@ -1,104 +1,457 @@
+            features["group"].nunique(),
+        )
+    # Define the mokapot model.
+    #   - Choice between a random forest, linear SVM, or no semi-supervised
+    #     learning.
+    #   - Features are preprocessed by standardizing them, removing
+    #     zero-variance features, and removing highly correlated features.
+    #   - We perform minimal tuning of the class weights and the random forest
+    #     tree depth.
+    if model is None:
+        logging.debug("Calculate q-values based on the cosine similarity")
+        # Calculate q-values based on the cosine similarity.
+        confidences = dataset.assign_confidence(features["cosine"], True)
+    else:
+        logging.debug(
+            "Train semi-supervised %s model and score SSMs", model.upper()
+        )
+        scaler = make_pipeline(
+            StandardScaler(),
+            VarianceThreshold(),
+            CorrelationThreshold(0.95),
+        )
+        if model == "svm":
+            clf = mokapot.model.PercolatorModel(
+                scaler, train_fdr=fdr, direction="cosine"
+            )
+        elif model == "rf":
+            clf = mokapot.Model(
+                GridSearchCV(
+                    RandomForestClassifier(random_state=1),
+                    param_grid={
+                        "max_depth": [3, 5, 7, 9, None],
+                        "class_weight": [
+                            None,
+                            {0: 0.1, 1: 1},
+                            {0: 0.1, 1: 10},
+                            {0: 1, 1: 0.1},
+                            {0: 1, 1: 10},
+                            {0: 10, 1: 0.1},
+                            {0: 10, 1: 1},
+                        ],
+                    },
+                    refit=False,
+                    cv=3,
+                    n_jobs=-1,
+                ),
+                scaler,
+                train_fdr=fdr,
+                direction="cosine",
+            )
         else:
-            break
+            raise ValueError(
+                "Unknown semi-supervised machine learning model given"
+            )
+        # Train the mokapot model and combine the SSMs for all groups.
+        confidences, _ = mokapot.brew(dataset, clf, fdr)
+    ssm_scores = pd.concat(
+        [
+            confidences.group_confidence_estimates[group].psms
+            for group in confidences.groups
+        ],
+        ignore_index=True,
+    )
+    # Assign the scores to the SSM objects.
+    for i, score, q in zip(
+        ssm_scores["index"],
+        ssm_scores["mokapot score"],
+        ssm_scores["mokapot q-value"],
+    ):
+        ssms[i].search_engine_score = score
+        ssms[i].q = q
+    return ssms
 
 
-def filter_group_fdr(ssms: Iterator[SpectrumSpectrumMatch], fdr: float = 0.01,
-                     tol_mass: float = None, tol_mode: str = None,
-                     min_group_size: int = None)\
-        -> Iterator[SpectrumSpectrumMatch]:
+def _get_ssm_groups(
```

Comment:
**question:** I've thought about adding this functionality in mokapot, but I'm not sure if its worth it. What do you think?

Reply:
To my knowledge there's no "default" way of grouping PSMs that everyone does, so there's no generally accepted way that mokapot would do it. Some custom implementation could be useful, but I don't consider it core functionality that's absolutely necessary.

Coincidentally this is very similar to how the PTM-Shepherd paper describes that they do it though, even though it was developed independently for another project without having read that paper. It would be useful to further validate this part of my code with unit tests on a representative example though to be sure it fully works as expected.